### ISSUE TITLE
[Suggestion] Class diagram in docs

### ISSUE THREAD

**Issue by jotaf98** [pyqtgraph/pyqtgraph#848]
Hi, I've been using PyQtGraph for a while and it's great!

Of course, it has a deep/interwoven hierarchy of classes (important for reusability and flexibility). Sometimes it is not absolutely clear how to put them together when you're building things from scratch (as opposed to just copying and modifying examples!).

This documentation page is immensely helpful, but only offers a partial view, focused on the graphics layout and a handful of classes:

http://www.pyqtgraph.org/documentation/plotting.html#organization-of-plotting-classes

Many classes and ways to compose things are missing. Would it be possible to complement this page with a complete diagram of the class hierarchy? Especially relationships like "can be a child of" are super useful when creating new graphics from scratch, and to have a "big picture" of sorts. I would offer to do it but my knowledge of the full class hierarchy is piecemeal and limited.

**Comment by j9ac9k** [comment]
We would certainly welcome and review a PR with such a change to the documentation.  Right now the main development focus is getting the next stable release out with official PySide2 support, so I doubt we'll create this ourselves.

**Comment by titusjan** [comment]
I wanted to make the same suggestion but saw that this was already reported by @jotaf98 .

The [organization-of-plotting-classes](https://pyqtgraph.readthedocs.io/en/latest/plotting.html#organization-of-plotting-classes) page gives a good start but, when I just started learning PyQtGraph, I struggled to get a good mental image of how the different classes related. I therefore made the following UML diagram for myself...

![uml](https://user-images.githubusercontent.com/3320415/110247534-fd6e6c80-7f6c-11eb-992d-590479b6f196.png)

It is made in [Dia](http://dia-installer.de/shapes/UML/index.html.en). It's not complete but I think the most important classes are there.

I'm willing to make a PR but I would like somebody to check it for correctness. Also, in which page should I add the diagram then? I don't know if the image will be readable if I just add it to that page.

**Comment by j9ac9k** [comment]
my goodness, I know our hierarchy is confusing ...and I often have to sketch it out myself, but wow, this is pretty brutal haha.  

read the docs actually supports the use of sphinx-plantuml so we should probably should just utilize that and have the diagram code actually be part of the repo instead of attaching an image itself.

https://github.com/readthedocs/readthedocs.org/issues/3885

I think trying to cram all the major pyqtgraph classes into one diagram will be pretty tough, and I don't think it will serve its intended purpose.  What we should probably do is create a class diagram for GraphicsItem, GraphicsWidget, GraphicsView and how they relate to one another (and what Qt objects they inherit); and then on the docs page for each of the more specific types (`PlotDataItem` and so on) we should a diagram of just its parents.

@ixjlyons I know you've spent some time thinking about this; what do you think?

**Comment by jotaf98** [comment]
Really good job! Any way to attach a legend (for the notation) there?

I have to say that a broken-down diagram with incomplete information wouldn't improve things much. A diagram with the parents only for each class, in particular, wouldn't convey any info that wasn't clear with just text (i.e. say in each class doc what are its parents). And we'd be back to the same situation as now.

No matter how complicated the diagram is, it reflects reality. No need to hide it! It is what it is. A global view is exactly what is needed to make the hierarchy clear in my mind, which is why @titusjan's image helps so much.

I guess you could also have a "simplified" diagram for non-advanced use, but that's what we already have [here](https://pyqtgraph.readthedocs.io/en/latest/plotting.html#organization-of-plotting-classes).

**Comment by ixjlyons** [comment]
I'll try to come back and give a more complete thought but my first impression is basically any complete or auto-generated diagram is good for reference but not great for learning. The main thing I'd like to see is a a concise description of the core classes like `GraphicsObject` and `GraphicsItem` and how they relate to the Qt classes so you can figure out what behaviors are being inherited where. The diagram above is a little overwhelming but is a very nice reference. I'd be in favor of having something like that in the docs but I'd also like to think about how to present the info with prose + simplified diagrams so it's easier to get acquainted for newcomers.

**Comment by titusjan** [comment]
Obviously I'm on the same page as @jotaf98; I think a complete diagram is essential to get a good understanding of the PyQtGraph library. 

Just to give a few examples:

* It makes it obvious that`PlotItem` and `PlotDataItem` are two totally different things, even though they both indirectly inherit from GraphicsItem and QGraphicsObject. Two separate diagrams won't convey that message IMO.
* It shows the distinct position of the ImageView class. It is really high in the class hierarchy.
* It makes it clear that only `PlotItem` and `ImageView` objects have a `ViewBox`.
* It demonstrates how to get from a `GraphicsLayoutWidget` to its `QGraphicsGridLayout`. 
 
I couldn't have answered SO questions like [this](https://stackoverflow.com/questions/45879148/pyqtgraph-help-understanding-source-code/45938026#45938026) and [this](https://stackoverflow.com/questions/36889159/pyqtgraph-graphics-layout-widget-issue/36897295#36897295) without my UML schema.

>  any complete or auto-generated diagram is good for reference but not great for learning. 

I totally agree. The complete diagram can be quite overwhelming for newcomers. It's more intended as a reference. Therefore, it should probably **not** be included in the [Organization of Plotting Classes](https://pyqtgraph.readthedocs.io/en/latest/plotting.html#organization-of-plotting-classes) page, but on a separate page in the API reference section.

**Comment by titusjan** [comment]
> read the docs actually supports the use of sphinx-plantuml so we should probably should just utilize that and have the diagram code actually be part of the repo instead of attaching an image itself.

I recreated the diagram in [PlantUML](https://plantuml.com/). It has the same classes and relations as the UML diagram above, except for two modifications:

1. I added an aggregation line from the `ImageView`class to the `ImageItem` class. 
2. I corrected a mistake. The `GraphicsWidget` class does not have a `layout` attribute. Instead this attribute is present in the `PlotItem` and `GraphicsLayout` classes.


> Any way to attach a legend (for the notation) there?

I'm afraid that a legend that says that a line with diamond means _aggregation_ doesn't help much. There still would be additional text necessary. Or a link to the UML wikipedia page. In any case, I think the new diagram is clearer because it shows the attributes in the classes and the diamond lines are connected those attributes.

I think the new diagram is an improvement. I created a pull request (#1631) so you guys can give feedback. 

![uml_overview 06-commit-865991f](https://user-images.githubusercontent.com/3320415/111039701-b6cab780-842f-11eb-8068-e3a681968290.png)
